### ISSUE TITLE
DeepMerge transforms Map to [key, value] Array in some conflicts

### ISSUE THREAD

**Issue by Methuselah96** [immutable-js-oss/immutable-js#70]
*From @CharlieScarver on Mon, 15 Jan 2018 16:47:59 GMT*

Hello.
Using ```immutable@4.0.0-rc.9```
I was doing a ```.deepMerge``` and I ran into a conflict between a List and Map like this one: 
```javascript
const a = Immutable.fromJS({
    ch: [{
        code: 8
    }]
}); 
const b = Immutable.fromJS({
    ch: {
        code: 8
    }
});
const result = a.mergeDeep(b); 
/* result:
Map {
    'ch': List[
        Map { 'code': 8 },
        Array [ 'code', 8 ]
    ]
}
*/
```
Then I tried this:
```javascript
const a = Immutable.fromJS({
    ch: []
}); 
const b = Immutable.fromJS({
    ch: { code: 8 }
});
const result = a.mergeDeep(b); 
/* result:
Map {
    'ch': List[
        Array [ 'code', 8 ]
    ]
}
*/
```
I guess there's a problem when merging a key with a List value and a key with a Map value (the Map has 
 to have at least one key inside).
So I'm baffled by that Array there. 
Is this a bug?

Thank you.

*Copied from original issue: https://github.com/immutable-js/immutable-js/issues/1475*

**Comment by Methuselah96** [comment]
Not sure how to categorize this. It's a change from 3.x, but I'm not sure that it's a bug since the intended behavior is not clear to me.

**Comment by Methuselah96** [comment]
I don't think this is defined behavior, so I'm going to take it off of the 4.0 milestone for now.

**Comment by CharlieScarver** [comment]
I believe leaving it as it is would be the worst option. Returning a regular mutable array will not be an expected behaviour in any code using Immutable.js structures.

Is there a reason why this case cannot be defined as the following:
When there is a conflict of two different structures take the value of B.

This would work like the other conflict cases:
```javascript
const a = Immutable.fromJS({
    code: 8
}); 
const b = Immutable.fromJS({
    code: "name"
});
const result = a.mergeDeep(b); 
/* result:
Map {
    code: "name"
}
*/
```

**Comment by Methuselah96** [comment]
I'll add it back to the 4.0 milestone to consider what we should be doing. Thanks!

**Comment by jdeniau** [comment]
I think that we can reduce this issue to this minimal  reproductible code:

https://runkit.com/jdeniau/immutable-merge-with-different-types

TLDR; 
Immutable 3.8 was already doing arrays instead of Lists.

Immutable 3.8 was weird because it did remove the item in the original List.
Immutable 4 is weird because it convert the Map to an Array.

My personnal preference, would be to throw an Error in that case, but I don't know if it is the immutable way of doing. This way the user understand really that he is doing something weird. But it might break a lot of existing code.

On the other hand, as the merge does take the second value if keys are identical, I would agree with @CharlieScarver to take the second value and drop the first if the type are different. It could be argued though because some type might be merged (List + Array or Map + Set + Record, etc.).

**Comment by Methuselah96** [comment]
See conversation in https://github.com/immutable-js-oss/immutable-js/pull/215.
